pkg/plugins/gateway/algorithms/vtc/token_tracker.go

```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2024 The Aibrix Team.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vtc
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/vllm-project/aibrix/pkg/utils"
+)
+
+// Sliding window configuration (minutes)
+const (
+	VTC_TOKEN_TRACKER_WINDOW_MINUTES = "AIBRIX_ROUTER_VTC_TOKEN_TRACKER_WINDOW_MINUTES"
+	defaultTokenTrackerWindowMinutes = 60
+)
+
+var (
+	tokenTrackerWindowMinutes = utils.LoadEnvInt(VTC_TOKEN_TRACKER_WINDOW_MINUTES, defaultTokenTrackerWindowMinutes)
+)
+
+type TimeUnit int
+
+const (
+	Minutes TimeUnit = iota
+	Seconds
+	Milliseconds
+)
+
+// Time duration mapping for each time unit
+var timeUnitDuration = map[TimeUnit]time.Duration{
+	Minutes:      time.Minute,
+	Seconds:      time.Second,
+	Milliseconds: time.Millisecond,
+}
+
+// Helper functions for time unit operations
+func (unit TimeUnit) truncateTime(t time.Time) time.Time {
+	return t.Truncate(timeUnitDuration[unit])
+}
+
+func (unit TimeUnit) toTimestamp(t time.Time) int64 {
+	if unit == Milliseconds {
+		return t.UnixNano() / int64(time.Millisecond)
+	}
+	return t.Unix()
+}
+
+// InMemorySlidingWindowTokenTracker tracks tokens per user in a fixed-size sliding window (in-memory, thread-safe).
+type InMemorySlidingWindowTokenTracker struct {
+	mu          sync.RWMutex
+	windowSize  time.Duration
+	buckets     int
+	bucketUnit  TimeUnit
+	userBuckets map[string]map[int64]float64 // user -> windowStart -> tokenSum
+	config      *VTCConfig
+}
+
+// TokenTrackerOption is a function that configures a token tracker
+type TokenTrackerOption func(*InMemorySlidingWindowTokenTracker)
+
+func WithWindowSize(size int) TokenTrackerOption {
+	return func(t *InMemorySlidingWindowTokenTracker) {
+		t.buckets = size
+		t.windowSize = time.Duration(size) * timeUnitDuration[t.bucketUnit]
+	}
+}
+
+func WithTimeUnit(unit TimeUnit) TokenTrackerOption {
+	return func(t *InMemorySlidingWindowTokenTracker) {
+		t.bucketUnit = unit
+		t.windowSize = time.Duration(t.buckets) * timeUnitDuration[unit]
```

Comment:
Seems `t.windowSize` is updated in both `WithWindowSize` and `WithTimeUnit`. and these two are being used together. then does the sequence matter?

Reply:
nice catch ! oversight on my side, fixed it.